nixos/modules/services/security/oauth2_proxy_nginx.nix

```
@@ -13,54 +17,66 @@ in
         The address of the reverse proxy endpoint for oauth2_proxy
       '';
     };
-    virtualHosts = mkOption {
-      type = types.listOf types.str;
-      default = [];
-      description = lib.mdDoc ''
-        A list of nginx virtual hosts to put behind the oauth2 proxy
+    virtualHostLocations = mkOption {
+      type = types.attrsOf (types.listOf types.str);
+      example = literalExpression ''
+        {
+          "example.com" = [ "/" "/assets/" ];
+        };
       '';
+      default = { };
+      description = lib.mdDoc "List of locations per virtual host to put behind the oauth2_proxy";
     };
   };
-  config.services.oauth2_proxy = mkIf (cfg.virtualHosts != [] && (hasPrefix "127.0.0.1:" cfg.proxy)) {
-    enable = true;
-  };
-  config.services.nginx = mkIf config.services.oauth2_proxy.enable (mkMerge
-  ((optional (cfg.virtualHosts != []) {
-    recommendedProxySettings = true; # needed because duplicate headers
-  }) ++ (map (vhost: {
-    virtualHosts.${vhost} = {
-      locations."/oauth2/" = {
-        proxyPass = cfg.proxy;
-        extraConfig = ''
-          proxy_set_header X-Scheme                $scheme;
-          proxy_set_header X-Auth-Request-Redirect $scheme://$host$request_uri;
-        '';
-      };
-      locations."/oauth2/auth" = {
-        proxyPass = cfg.proxy;
-        extraConfig = ''
-          proxy_set_header X-Scheme         $scheme;
-          # nginx auth_request includes headers but not body
-          proxy_set_header Content-Length   "";
-          proxy_pass_request_body           off;
-        '';
-      };
-      locations."/".extraConfig = ''
-        auth_request /oauth2/auth;
-        error_page 401 = /oauth2/sign_in;
+  config = {
+    assertions = [{
+      assertion = !builtins.any (loc: builtins.elem loc (flatten (builtins.attrValues cfg.virtualHostLocations))) [ "/oauth2/" "/oauth2/auth" ];
```

```suggestion
      assertion = !lib.any (loc: lib.elem loc (flatten (lib.attrValues cfg.virtualHostLocations))) [ "/oauth2/" "/oauth2/auth" ];
```
if there is a lib counterpart of a lib function that one should be used.